src/main/java/net/lingala/zip4j/headers/HeaderWriter.java

```diff
@@ -455,7 +459,9 @@ private void writeFileHeader(ZipModel zipModel, FileHeader fileHeader, ByteArray
       byte[] fileCommentBytes = new byte[0];
       if (isStringNotNullAndNotEmpty(fileComment)) {
         Charset charset = Charset.forName(InternalZipConstants.ZIP_STANDARD_CHARSET);
-        if (BitUtils.isBitSet(fileHeader.getGeneralPurposeFlag()[1], 3)) {
+        if(HeaderUtil.isCharsetValid(charsetName)) {
+          charset = Charset.forName(charsetName);
+        } else if (BitUtils.isBitSet(fileHeader.getGeneralPurposeFlag()[1], 3)) {
```

Comment:
Once you change the charset to Charset from String, this if-else condition will change accordingly. Since we will always have a default charset, there will not be an if-else here.